### DISCUSSION TITLE
feat: support up to 16 cities (split in half: limit by rbgen + urban planning centers)

### DISCUSSION THREAD

**Discussion by winex** [bahrmichael/factorio-tycoon#276] [Ideas]
_ok, as it's very technical idea i'm using github instead of mod portal for such discussion_

### proposal
increase number of towns to **16**, but allow only 8+8:
- 8 generated ones (will add a limit in rbgen)
- 8 ~~by UPC or~~ w/e-placed, _like reserved for future use or be freed then_

16 is because it's power-of-two (halves down to 1 easily, `sqrt()`'d only to 2). also, there will be not that many players who will try to pass this limit (unlike us ;) ), because
- to have 8 towns by rbgen w/ 1024 (current) region size a map should be `1024 tiles * sqrt(8 towns)/(0.5 prob) =` **~6Kx6K**, which is pretty big 

~~if you ever decide to remove UPC or manual placement at all~~ -- we'll just add that freed 8 slots into `rbgen` limit, that's it.

### also planning:
- [ ] fix: random name choosing for towns from **full list** (enable it), but avoid duplicates by remembering used ones
why initial city is always `Ironhaven`?!
- [x] rbgen: add limit for towns
- [x] rbgen: forbid spawning more than 1 town per region
- [ ] rbgen: deprecate manual hack called `Spawn Initial City` -- this should be removed
- ??? remove town name from passenger prototype, replace that with `"-" .. N` suffix -- need to find some way to add that by `cityId` in gui, if that's possible.
- or just call them `Passenger 1`, `Passenger 2` and write side-info table of current name assignments somewhere in stations gui
- maybe this is related info: https://forums.factorio.com/viewtopic.php?t=46687
- found some more: `custom_description` field for `item-with-tags`, this looks very useful: https://forums.factorio.com/viewtopic.php?t=77569

### passenger trick
yeah, i know about that - like that a lot! as Factorio weren't designed for `cargodist` (as it is named in `OpenTTD`), i also think about exploring some possibilities like using item's HP to encode destination (though inserter filters seems not to support it) or maybe some custom field if that's possible

### spamming prototypes
about items: 6, 12 or 16 doesn't differ much, but way less than 100, which could lag Helmod or some others.
some time ago i was hacking on `alien-economy-light` (it was adding 100 recipes of `alien-hyper-module`s) -- this was freaking lagging most of recipe-browsing mods like FNEI, Helmod, RecipeBook(?), especially in multiplayer (on LAN!) ...until i decreased it by a factor of 5 - it's just 20 items now by +5 level each.

**Comment by bahrmichael** [comment]
> why initial city is always Ironhaven

That's coming from https://github.com/bahrmichael/factorio-tycoon/blob/49c2789f9e55dd1b441b868c45d08d4e4d3f4880/city-planner.lua#L341. Looking at the list of possible names it could've started one earlier, but I'm fine with it. We can use a randomizer there if we want.

**Comment by bahrmichael** [comment]
> ??? remove town name from passenger prototype, replace that with "-" .. N suffix -- need to find some way to add that by cityId in gui, if that's possible.

The reason I went with "Passenger to Ironhaven" and the others, is that you can set Factorio filters (e.g. on inserters) to those dedicated items. As long as we can maintain that concept or find a good alternative, I'm happy to change it up.

The main idea here was that you could have larger passenger hubs where they're sorted onto smaller trains. Think long-distance trains and local trains.

**Comment by bahrmichael** [comment]
> found some more: custom_description field for item-with-tags, this looks very useful: https://forums.factorio.com/viewtopic.php?t=77569

If you need inspiration: We're already using item-with-tags to remember when a passenger started to caculate the reward.

**Comment by bahrmichael** [comment]
> increase number of towns to 16, but allow only 8+8:
> 8 generated ones (will add a limit in rbgen)
> 8 by UPC or w/e-placed, like reserved for future use or be freed then

I'm cool with that :)
